src/clj/puppetlabs/services/jruby_pool_manager/impl/reference_pool.clj

```diff
@@ -41,15 +43,21 @@
   [pool-context :- jruby-schemas/PoolContext
    instance :- JRubyInstance]
   (let [borrow-count (:borrow-count pool-context)
-        max-borrows (get-in instance [:internal :max-borrows])]
-    (pool-protocol/lock pool-context)
+        max-borrows (get-in instance [:internal :max-borrows])
+        flush-timeout (jruby-internal/get-flush-timeout pool-context)]
     (try
-      ;; Now that we've successfully acquired the lock, check the borrows again
-      ;; to make sure the pool wasn't flushed while we were waiting.
-      (when (max-borrows-exceeded @borrow-count max-borrows)
-        (flush-pool* pool-context))
-      (finally
-        (pool-protocol/unlock pool-context)))))
+      ;; Lock will block until all references have been returned to the pool or
+      ;; until flush-timeout is reached
+      (pool-protocol/lock-with-timeout pool-context flush-timeout TimeUnit/MILLISECONDS)
+      (try
+        ;; Now that we've successfully acquired the lock, check the borrows again
+        ;; to make sure the pool wasn't flushed while we were waiting.
+        (when (max-borrows-exceeded @borrow-count max-borrows)
+          (flush-pool* pool-context))
+        (finally
+          (pool-protocol/unlock pool-context)))
+      (catch TimeoutException e
+        (log/info (i18n/trs "An attempt to lock the JRubyPool failed with a timeout"))))))
```

Comment:
Can we put some context here about why we were trying to lock? Maybe saying something about how we were trying to flush because we'd hit the max borrows per instance, and we'll try again?

Reply:
For sure. I was trying to keep the message the same as the one for the other pool, but I definitely agree that more info would be useful.